dali/plugin/copy.cu

```diff
@@ -22,51 +22,74 @@
 
 namespace dali {
 
-void CopyToExternalTensor(const Tensor<CPUBackend>& t, void* ptr,
-                                            device_type_t dst_type) {
-  DALI_ENFORCE(t.ndim() > 0, "Can't copy empty Tensor!");
+template <typename T>
+static void CopyToExternalTensorHelper(const dali::Buffer<T> &src, void *dst,
+                                device_type_t dst_type, size_t num) {}
+
+template <>
+void CopyToExternalTensorHelper<CPUBackend>(const dali::Buffer<CPUBackend> &src, void *dst,
+                                            device_type_t dst_type, size_t num) {
   if (dst_type == CPU) {
-    std::memcpy(ptr,
-              t.raw_data(),
-              Volume(t.shape()) * t.type().size());
+    std::memcpy(dst, src.raw_data(), num);
   } else {
     DALI_FAIL("Coping from CPUBackend to device type " + to_string(dst_type));
   }
 }
 
-void CopyToExternalTensor(const Tensor<GPUBackend>& t, void* ptr,
-                                            device_type_t dst_type) {
-  DALI_ENFORCE(t.ndim() > 0, "Can't copy empty Tensor!");
-  DeviceGuard d(t.device_id());
-  cudaStream_t stream = UserStream::Get()->GetStream(t);
+template <>
+void CopyToExternalTensorHelper<GPUBackend>(const dali::Buffer<GPUBackend> &src, void *dst,
+                                            device_type_t dst_type, size_t num) {
+  DeviceGuard d(src.device_id());
   cudaMemcpyKind direction;
+  cudaStream_t stream = UserStream::Get()->GetStream(src);
   if (dst_type == GPU) {
     direction = cudaMemcpyDeviceToDevice;
   } else if (dst_type == CPU) {
     direction = cudaMemcpyDeviceToHost;
   } else {
     DALI_FAIL("Coping from GPUBackend to device type " + to_string(dst_type));
   }
-  CUDA_CALL(cudaMemcpyAsync(ptr,
-                            t.raw_data(),
-                            Volume(t.shape()) * t.type().size(),
+  CUDA_CALL(cudaMemcpyAsync(dst,
+                            src.raw_data(),
+                            num,
                             direction,
                             stream));
   CUDA_CALL(cudaStreamSynchronize(stream));
 }
 
+template <typename T>
+static void CopyToExternalTensorListHelper(TensorList<T>* tl, void* ptr,
+                                       device_type_t dst_type) {
+  if (tl->IsDenseTensor()) {
+    Tensor<T> t;
+    t.ShareData(tl);
+    CopyToExternalTensor(t, ptr, dst_type);
+  } else {
+    CopyToExternalTensorHelper<T>(*tl, ptr, dst_type, tl->nbytes());
+  }
+}
+
+void CopyToExternalTensor(const Tensor<CPUBackend>& t, void* ptr,
+                                            device_type_t dst_type) {
```

Comment:
indentation. I am thinking that the linter could/should help with those things

Reply:
Done